Implement custom label header support for Grpc metrics interceptor

rastenis opened the pull request:
Add an option to provide a list of custom label headers to add to metrics.

werkt commented on src/main/java/build/buildfarm/common/config/GrpcMetrics.java:
```diff
@@ -10,6 +10,7 @@ public class GrpcMetrics {
   private boolean enabled = false;
   private boolean provideLatencyHistograms = false;
   private double[] latencyBuckets;
+  private String labelsToReport = "";
```
Make this an array, no need to split based on undocumented comma separator.

rastenis replied:
Switched it to a List, since that is what `grpcConfig.withLabelHeaders` takes. Let me know if an Array field with a Array->List conversion at the time of calling is preferred.

The reason this variable was a string previously was due to the variable being loaded through an environmental variable in production, which can only be a string without additional conversion. Where can I ensure that `BuildfarmConfigs` contains the correct format of `labelsToReport`?

werkt reviewed (changes requested):
Please detail some more motivation for this change in the description, some 'why' as to the usefulness of this change.
Also we're missing some additional requirements here for new config options:
1. Example in examples/config.yml
2. Documentation in _site/configuration/configuration.md

rastenis commented:
> Please detail some more motivation for this change in the description, some 'why' as to the usefulness of this change. Also we're missing some additional requirements here for new config options:
> 
> 1. Example in examples/config.yml
> 2. Documentation in _site/configuration/configuration.md

1. Added an example. By default, the labelsToReport should be empty, since it adds cardinality to metrics. 
2. Added documentation

werkt commented on src/main/java/build/buildfarm/common/config/GrpcMetrics.java:
```diff
@@ -30,6 +32,11 @@ public static void handleGrpcMetricIntercepts(
         grpcConfig = grpcConfig.withLatencyBuckets(grpcMetrics.getLatencyBuckets());
       }
 
+      // report custom metric labels
+      if (grpcMetrics.getLabelsToReport() != null && !grpcMetrics.getLabelsToReport().isEmpty()) {
+        grpcConfig = grpcConfig.withLabelHeaders(grpcMetrics.getLabelsToReport());
+      }
+
```
Shouldn't need the empty check here - we don't do it on the latency buckets above.

rastenis replied:
Removed. I put the check there to avoid unknown behaviour from passing empty label arrays to grpcConfig.withLabelHeaders, since I did not look at the implementation of that function to ensure it is not an issue.

werkt replied:
I took a look at it. Good consideration, although the prior art above with existing list specification, and general API expectations are that adding all elements of an empty list should noop and represent no difference to not calling the add in the first place.

amishra-u commented on examples/config.yml:
```diff
@@ -13,6 +13,7 @@ server:
     enabled: true
     provideLatencyHistograms: true
     latencyBuckets: [0.001, 0.01, 0.1, 1, 5, 10, 20, 40, 60, +Infinity]
+    labelsToReport: []
```
You might want to add it under worker config also. line #99.

amishra-u commented:
> Please detail some more motivation for this change in the description, some 'why' as to the usefulness of this change. Also we're missing some additional requirements here for new config options:
> 
> 1. Example in examples/config.yml
> 2. Documentation in _site/configuration/configuration.md

We want to distinguish the metrics between cache warmup jobs and requests from developer machines. The[ java-grpc-prometheus](https://github.com/grpc-ecosystem/java-grpc-prometheus/blob/master/src/main/java/me/dinowernli/grpc/prometheus/Labels.java#L32) library already supports attaching any configured key from RequestMetadata to the Collector.  We will make some changes client side so that RemoteRequests for cache warmup job and requests from developer machines will carry different values for a custom key in RequestMetadata.
